Provide improved generation of arbitrary `UTCTime` values.

jonathanknowles opened the pull request:
# Issue Number

https://github.com/input-output-hk/cardano-wallet/wiki/Blackboard#investigate-the-arbitrary-instance-for-utctime-in-testquickcheckinstancestime

# Overview

This change provides a `newtype` wrapper for `UTCTime` whose `Arbitrary` instance spans a uniform range of dates and a mixture of time precisions (_hour_, _second_ and _picosecond_ precision).

# Examples

### Generation
```hs
% replicateM_ 10 (generate arbitrary >>= print . getUniformTime)
1986-11-30 00:00:00 UTC
2471-07-30 02:04:41 UTC
2534-08-23 15:01:12.29087396788 UTC
2007-05-23 04:12:14.161001565693 UTC
2778-01-30 18:27:57 UTC
2025-09-05 12:00:00 UTC
2266-03-11 17:20:21 UTC
2822-08-14 21:40:18 UTC
2319-05-18 22:00:00 UTC
2448-03-09 11:38:11.71875851142 UTC
```

KtorZ commented on lib/core/src/Test/QuickCheck/Instances/UTCTime.hs:
```diff
@@ -0,0 +1,67 @@
+            | x <- shrink $ toModifiedJulianDay numberOfDays ]
+        , [ t { utctDayTime = picosecondsToDiffTime y }
+            | y <- shrink $ diffTimeToPicoseconds timeSinceMidnight ]
+        ]
```
Not sure how relevant it is to shrink an arbitrary time ... :thinking: 
I don't think it helps getting a _simpler_ value in anyway :sweat_smile: ? It justs get another arbitrary time and burns CPU cycles, no?

jonathanknowles replied:
> Not sure how relevant it is to shrink an arbitrary time ... thinking
> I don't think it helps getting a _simpler_ value in anyway sweat_smile ? It justs get another arbitrary time and burns CPU cycles, no?

Fair enough! My original thinking was to:

1. _Improve upon_ the _problematic_ behaviour of the original instance (generation).
2. _Preserve non-problematic behaviour_ as is (shrinking).

However, I also can't think of a useful use case for context-free shrinking. I'll remove the shrinker.

KtorZ commented on lib/core/src/Test/QuickCheck/Instances/UTCTime.hs:
```diff
@@ -0,0 +1,67 @@
+picosecondsInOneSecond :: Integral a => a
+picosecondsInOneSecond = 1_000_000_000_000
+
+-- | The maxmimum number of seconds in one day, allowing for leap seconds.
```
~~maxmimum~~ -> *maximum*

jonathanknowles replied:
Well spotted!

KtorZ commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -60,6 +60,7 @@ library
     , persistent
     , persistent-sqlite
     , persistent-template
+    , QuickCheck
```
That is very regrettable though :s ... We kept our core code free of any test assets so far an, I admit that introducing a QC dependency at this level makes me particularly sad / anxious.

jonathanknowles replied:
This is a fair point.

I've moved the instance to a separate package `test-utils` that can be imported just by targets related to testing.

KtorZ commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -108,6 +109,7 @@ library
       Data.Quantity
       Network.Wai.Middleware.ServantError
       Network.Wai.Middleware.Logging
+      Test.QuickCheck.Instances.UTCTime
```
I'd rather have that in a shared test module :s

jonathanknowles replied:
Done. I've moved the instance (and `newtype` wrapper) to a separate package `test-utils` that can be imported just by targets related to testing.

KtorZ commented on lib/core/src/Test/QuickCheck/Instances/UTCTime.hs:
```diff
@@ -0,0 +1,67 @@
+
+-- | An `Arbitrary` instance for `UTCTime` that spans a wide range of dates
+--   and a mixture of time precisions.
+instance Arbitrary UTCTime where
```
Couldn't a newtype (e.g. `UniformUTCTime ...`) help both:

- Conveying the meaning behind this instance (more uniformly distributed than the default arbitrary)
- Avoiding the orphan instance which makes imports quite unclear

Perhaps even sticking with `genUniformUTCTime :: Gen UTCTime` since the shrinker isn't really of any use for this one. And, import this helper wherever we need to define a uniform UTCTime.

jonathanknowles replied:
I agree. I'll update the PR so that we define a `newtype` wrapper (as suggested), and make an instance for the `newtype` instead of for `UTCTime` directly.

jonathanknowles committed:
Add package `test-utils` with utilities for writing unit and property tests.

To start with, this package provides a newtype wrapper for `UTCTime` whose
`Arbitrary` instance spans a uniform range of dates and a mixture of time
precisions.

It also provides a standalone generation function `genUniformTime`.

Examples:

```
% replicateM_ 10 (generate arbitrary >>= print . getUniformTime)
1986-11-30 00:00:00 UTC
2471-07-30 02:04:41 UTC
2534-08-23 15:01:12.29087396788 UTC
2007-05-23 04:12:14.161001565693 UTC
2778-01-30 18:27:57 UTC
2025-09-05 12:00:00 UTC
2266-03-11 17:20:21 UTC
2822-08-14 21:40:18 UTC
2319-05-18 22:00:00 UTC
2448-03-09 11:38:11.71875851142 UTC
```

jonathanknowles committed:
Use `genUniformTime` in module `Cardano.Wallet.Api.TypesSpec`.

jonathanknowles committed:
Use `genUniformTime` in module `Cardano.Wallet.DBSpec`.

jonathanknowles committed:
Use `getUniformTime` in module `Data.Time.TextSpec`.
